### ISSUE TITLE
Remove the `no-url-literals` experimental feature

### ISSUE THREAD

**Issue by thufschmitt** [NixOS/nix#10048]
**Is your feature request related to a problem? Please describe.**

Nix currently has a [`no-url-literals`](https://nixos.org/manual/nix/stable/contributing/experimental-features.html?highlight=url%20literal#no-url-literals) experimental feature that disables the use of URL literals in the language.
While the change by itself is something we agree we want to move towards, having it be an experimental feature isn't the right thing because it's unlikely that we'll ever flip the switch, and it isn't visible by default.

**Describe the solution you'd like**

What we should instead have is have it be (visibly) deprecated (through warnings), with a way to forbid it altogether, and a way to shut the warnings if needed.

-- _End of the team discussion, the below is my own interpretation of the agreement we had and up to discussion_ --

Concretely, we should remove the experimental feature, and add an option `warn-url-literal = {yes|no|error}`, defaulting to `yes`,  with the (long-term) plan of making it default to `error`.

This could be later extended to a more generic mechanism able to handle more warning/errors. Tvix has [a few of them](https://code.tvl.fyi/tree/tvix/eval/src/warnings.rs?id=6b136dfd23c03cc61d189ad5d41dd881e45b677e#n9), and there are a number of things discussed here that are blocked on the lack of a deprecation path (https://github.com/NixOS/nix/issues/7578, https://github.com/NixOS/nix/issues/8738, https://github.com/NixOS/nix/issues/8613)

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by edolstra** [comment]
Yeah, it should be generalized, e.g. `warnings = url-literals ...`. And then there could be an flag like gcc's `-Werror` to convert warnings into errors.

**Comment by roberth** [comment]
> e.g. `warnings = url-literals ...`.

What's nice about key/value, as it currently is, is that it's easy to override it. We have `extra-*` processing, but not `remove-`.
Similarly, when configured through the module system with RFC 42 `nix.settings`, the same dynamic applies.

**Comment by sellout** [comment]
The link to [no-url-literals](https://nix.dev/manual/nix/2.32/development/experimental-features.html#no-url-literals) in the description is broken, so here’s a more permanent URL.
